### PR TITLE
ci: enforce commit message convention, fixes #5037 [skip ci]

### LINKED ISSUE #5037: Use standardized PR titles, resulting in standardized master-branch commit messages
DDEV should also use some standards in PR titles and resulting merged commit messages. The commit message is generated from the PR title so the PR title has to follow the convention, we do not care about commit messages in a PR itself as it gets squash merged anyway. This has various advantages, e.g. we can find bugfixes to be back ported to a previous minor release. There is the widely used standard <https://www.conventionalcommits.org> which we can relay on.

This opens also new abillities e.g. automatically adding labels to PRs which later can be used to group the change log in the release by type.

## Definition of the commit message

`<type>[optional !]: <description>[, fixes #<issue>]`

This corresponds to the *Conventional Commits* and in addition the following rules must appy:

### Type
Must be one of the following:

* **build**: Changes that affect the build system or external dependencies
* **ci**: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
* **docs**: Documentation only changes
* **feat**: A new feature
* **fix**: A bug fix
* ***perf**: A code change that improves performance* (not sure if this has a value and is needed for DDEV at all)
* **refactor**: A code change that neither fixes a bug nor adds a feature
* **style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
* **test**: Adding missing tests or correcting existing tests

Regarding SemVer, all types above except `feat` increase the patch version, `feat` increases the minor version.

### Scope
No scope must be used.

### Breaking Changes
Breaking changes must have a `!` appended after type/scope.

Regarding SemVer, breaking changes increase the major version.

### Subject / Description
The subject contains a succinct description of the change:

* use the imperative, present tense: "change" not "changed" nor "changes"
* don't capitalize the first letter
* no dot (.) at the end

If an issue exists for the change, `, fixes #<issue number>` must be appended to the subject.

### Revert
If the commit reverts a previous commit, it should begin with `revert: `, followed by the header of the reverted commit. In the body it should say: `This reverts commit <hash>.`, where the hash is the SHA of the commit being reverted.

## Enforce the convention

The final definition will be enforced using https://github.com/marketplace/actions/gs-commit-message-checker, here is a prepared RegEx https://regex101.com/r/Dq4iGI/1

### PR DESCRIPTION
**Pull request by gilbertsoft**
* #5037 

<a href="https://gitpod.io/#https://github.com/ddev/ddev/pull/5038"></a>

### CODE REVIEW

**Comment by rfay** [comment]
Please make sure to lead the docs (and error messages from github actions) with several examples of valid titles. Most of us can't read through the requirements, we just want to do it right. Examples will help a lot.

File: .github/pr-labeler.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+labels:
+  - label: 'maintenance'
+    sync: true
+    matcher:
+      title: '^build(|!): .+'
+
+  - label: 'maintenance'
+    sync: true
+    matcher:
+      title: '^ci(|!): .+'
```
Comment (rfay) [inline .github/pr-labeler.yml:12]: You'll have to explain why there are two of these.

File: .github/workflows/pr-check.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+permissions:
+  contents: read
+  issues: write
+  pull-requests: write
+  statuses: write
+  checks: write
+      
+jobs:
+  check_commit_message:
+    name: Check Commit Message
```
Comment (rfay) [inline .github/workflows/pr-check.yml:16]: Looks like this is checking on ordinary commits? Is that true? We really only want it to check on master, as discussed.

File: docs/content/developers/building-contributing.md [hunk 3]
```diff
@@ -266,6 +266,64 @@ Once you’ve opened a pull request, a discussion will start around your propose
 
 If your pull request is merged, great! If not, no sweat; it may not be what the project maintainer had in mind, or they were already working on it. This happens, so our recommendation is to take any feedback you’ve received and go forth and pull request again. Or create your own open source project.
 
+### Pull Request Title Guidelines
+
+We have very precise rules over how our Git commit messages can be formatted. This leads to **more readable messages** that are easy to follow when looking through the **project history**. But also, we use the Git commit messages to **generate the changelog** for the releases.
```
Comment (rfay) [inline docs/content/developers/building-contributing.md:271]:
```suggestion
We have very precise rules over how our PR titles (and thus master-branch commits) are to be formatted. This leads to **more readable messages** that are easy to follow when looking through the **project history**. But also, we use the master-branch Git commit messages to **generate the changelog** for the releases.
```

File: docs/content/developers/building-contributing.md [hunk 4]
```diff
@@ -266,6 +266,64 @@ Once you’ve opened a pull request, a discussion will start around your propose
...
+* don't capitalize the first letter
+* no dot (.) at the end
+
+If an issue exists for the change, `, fixes #<issue number>` must be appended to the subject.
+
+#### Revert
+
+If the commit reverts a previous commit, it should begin with `revert:`, followed by the header of the reverted commit. In the body it should say: `This reverts commit <hash>.`, where the hash is the SHA of the commit being reverted.
+
+#### Examples
```
Comment (rfay) [inline docs/content/developers/building-contributing.md:317]: I think move this to the top, where people will see it right away, right before "Type".

File: docs/content/developers/building-contributing.md [hunk 5]
```diff
@@ -266,6 +266,64 @@ Once you’ve opened a pull request, a discussion will start around your propose
...
+`<type>[optional !]: <description>[, fixes #<issue>]`
+
+#### Type
+
+Must be one of the following:
+
+* **build**: Changes that affect the build or external dependencies
+* **ci**: Changes to our CI configuration files and scripts
+* **docs**: Documentation only changes
+* **feat**: A new feature
```
Comment (rfay) [inline docs/content/developers/building-contributing.md:284]: I see "enhancement" several places, is "feat" same as that?

**Review by rfay — CHANGES_REQUESTED** [review]
Looking good. I'm concerned that we haven't run this on master yet, have you run these on master on your fork? If not, please experiment a bit on your fork with these on master.

File: .github/workflows/pr-check.yml (hunk 2 above)
Comment (gilbertsoft) [reply .github/workflows/pr-check.yml:16]: That's a question of the configuration of the commit-message-checker bellow. Here it is configured to look at PR titles and body only. And as you can see above the event is pull_request_target (btw we need target here for the labeler or no labels could be added)

File: docs/content/developers/building-contributing.md (hunk 4 above)
Comment (gilbertsoft) [reply docs/content/developers/building-contributing.md:317]: done

File: docs/content/developers/building-contributing.md (hunk 3 above)
Comment (gilbertsoft) [reply docs/content/developers/building-contributing.md:271]: done

**Review by rfay — APPROVED** [review]
Sounds good, thanks!

Please make sure to test to your satisfaction on your fork before pulling.

**Comment by gilbertsoft** [comment]
Ready to be merged once @mattstein approved. Tested with Randy on my cloned repo and works like expected. Next step will be prepare release drafts based on the new labels.

**Pull request merged**